lib/interceptor/cache.js

```diff
@@ -0,0 +1,180 @@
+'use strict'
+
+const util = require('../core/util')
+const CacheHandler = require('../handler/cache-handler')
+const MemoryCacheStore = require('../cache/memory-cache-store')
+const CacheRevalidationHandler = require('../handler/cache-revalidation-handler')
+const { UNSAFE_METHODS } = require('../util/cache.js')
+
+const AGE_HEADER = Buffer.from('age')
+
+/**
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions | undefined} globalOpts
+ * @returns {import('../../types/dispatcher.d.ts').default.DispatcherComposeInterceptor}
+ */
+module.exports = globalOpts => {
+  if (!globalOpts) {
+    globalOpts = {}
+  }
+
+  if (globalOpts.store) {
+    for (const fn of ['createReadStream', 'createWriteStream', 'deleteByOrigin']) {
+      if (typeof globalOpts.store[fn] !== 'function') {
+        throw new TypeError(`CacheStore needs a \`${fn}()\` function`)
+      }
+    }
+
+    if (typeof globalOpts.store.isFull !== 'boolean') {
+      throw new TypeError(`CacheStore needs a isFull getter with type boolean, current type: ${typeof globalOpts.store.isFull}`)
+    }
+  } else {
+    globalOpts.store = new MemoryCacheStore()
+  }
+
+  if (globalOpts.methods) {
+    if (!Array.isArray(globalOpts.methods)) {
+      throw new TypeError(`methods needs to be an array, got ${typeof globalOpts.methods}`)
+    }
+
+    if (globalOpts.methods.length === 0) {
+      throw new Error('methods must have at least one method in it')
+    }
+  } else {
+    globalOpts.methods = ['GET']
+  }
+
+  // Safe methods the user wants and unsafe methods
+  const methods = [...globalOpts.methods, ...UNSAFE_METHODS]
+
+  return dispatch => {
+    return (opts, handler) => {
+      if (!opts.origin || !methods.includes(opts.method)) {
+        // Not a method we want to cache or we don't have the origin, skip
+        return dispatch(opts, handler)
+      }
+
+      const stream = globalOpts.store.createReadStream(opts)
+      if (!stream) {
+        // Request isn't cached
+        return dispatch(opts, new CacheHandler(globalOpts, opts, handler))
+      }
+
+      /**
+       * @param {import('../../types/cache-interceptor.d.ts').default.CacheStoreReadable | undefined} stream
+       * @param {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue} value
+       */
+      const respondWithCachedValue = (stream, value) => {
+        const ac = new AbortController()
+        const signal = ac.signal
+
+        signal.onabort = (_, err) => {
+          stream.destroy()
+          handler.onError(err)
+        }
+
+        stream.on('error', (err) => {
+          handler.onError(err)
```

Comment:
We don't want to call `onError` multiple times, e.g. if the `catch` handler has already called `onError` this should be a noop and vice-versa.